# Comparing `tmp/TileSeqMut-0.6.943-py3-none-any.whl.zip` & `tmp/TileSeqMut-0.6.944-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,28 +1,28 @@
-Zip file size: 62652 bytes, number of entries: 26
--rw-r--r--  2.0 unx      178 b- defN 22-May-25 18:10 TileSeqMut/__init__.py
--rw-r--r--  2.0 unx      119 b- defN 22-May-25 18:10 TileSeqMut/__main__.py
--rw-r--r--  2.0 unx     2094 b- defN 22-May-25 18:10 TileSeqMut/alignment.py
--rw-r--r--  2.0 unx    15800 b- defN 22-Oct-24 16:58 TileSeqMut/cluster.py
--rw-r--r--  2.0 unx    27653 b- defN 22-Jul-15 17:58 TileSeqMut/count_mut.py
--rw-r--r--  2.0 unx     4208 b- defN 22-May-25 18:10 TileSeqMut/help_functions.py
--rw-r--r--  2.0 unx    24363 b- defN 22-Jul-15 17:59 TileSeqMut/locate_mut.py
--rw-r--r--  2.0 unx    33636 b- defN 23-Apr-14 19:01 TileSeqMut/main.py
--rw-r--r--  2.0 unx    18853 b- defN 22-May-25 18:10 TileSeqMut/mergeruns.py
--rw-r--r--  2.0 unx    17020 b- defN 22-May-25 18:10 TileSeqMut/parse_clinvar.py
--rw-r--r--  2.0 unx    21518 b- defN 22-Jul-15 17:57 TileSeqMut/posterior.py
--rw-r--r--  2.0 unx    22185 b- defN 22-May-25 18:10 TileSeqMut/posterior_qc.py
--rw-r--r--  2.0 unx     1546 b- defN 22-May-25 18:10 TileSeqMut/process.py
--rw-r--r--  2.0 unx     3073 b- defN 22-May-25 18:10 TileSeqMut/random_ds.py
--rw-r--r--  2.0 unx     1747 b- defN 22-May-25 18:10 TileSeqMut/yogiroc_embeded.R
--rw-r--r--  2.0 unx     5470 b- defN 22-May-25 18:10 TileSeqMut/data/phix.fasta
--rwxr-xr-x  2.0 unx     2579 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.data/scripts/makePRC
--rwxr-xr-x  2.0 unx     1198 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.data/scripts/mergeRuns
--rwxr-xr-x  2.0 unx      661 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.data/scripts/posterior_QC
--rwxr-xr-x  2.0 unx      815 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.data/scripts/random_ds
--rwxr-xr-x  2.0 unx     4098 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.data/scripts/tileseq_mut
--rw-r--r--  2.0 unx     1048 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.dist-info/LICENSE
--rw-r--r--  2.0 unx    14351 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.dist-info/METADATA
--rw-r--r--  2.0 unx       92 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.dist-info/WHEEL
--rw-r--r--  2.0 unx       11 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.dist-info/top_level.txt
-?rw-rw-r--  2.0 unx     2199 b- defN 23-Apr-14 19:09 TileSeqMut-0.6.943.dist-info/RECORD
-26 files, 226515 bytes uncompressed, 59114 bytes compressed:  73.9%
+Zip file size: 62784 bytes, number of entries: 26
+-rw-rw-r--  2.0 unx      178 b- defN 23-Jul-18 13:45 TileSeqMut/__init__.py
+-rw-rw-r--  2.0 unx      119 b- defN 23-Jul-18 13:45 TileSeqMut/__main__.py
+-rw-rw-r--  2.0 unx     2094 b- defN 23-Jul-18 13:45 TileSeqMut/alignment.py
+-rw-rw-r--  2.0 unx    16108 b- defN 23-Aug-01 14:34 TileSeqMut/cluster.py
+-rw-rw-r--  2.0 unx    27641 b- defN 23-Aug-01 14:28 TileSeqMut/count_mut.py
+-rw-rw-r--  2.0 unx     4224 b- defN 23-Jul-18 13:45 TileSeqMut/help_functions.py
+-rw-rw-r--  2.0 unx    24363 b- defN 23-Jul-18 13:45 TileSeqMut/locate_mut.py
+-rw-rw-r--  2.0 unx    33680 b- defN 23-Aug-01 14:36 TileSeqMut/main.py
+-rw-rw-r--  2.0 unx    18853 b- defN 23-Jul-18 13:45 TileSeqMut/mergeruns.py
+-rw-rw-r--  2.0 unx    17020 b- defN 23-Jul-18 13:45 TileSeqMut/parse_clinvar.py
+-rw-rw-r--  2.0 unx    21518 b- defN 23-Jul-18 13:45 TileSeqMut/posterior.py
+-rw-rw-r--  2.0 unx    22185 b- defN 23-Jul-18 13:45 TileSeqMut/posterior_qc.py
+-rw-rw-r--  2.0 unx     1546 b- defN 23-Jul-18 13:45 TileSeqMut/process.py
+-rw-rw-r--  2.0 unx     3073 b- defN 23-Jul-18 13:45 TileSeqMut/random_ds.py
+-rw-rw-r--  2.0 unx     1747 b- defN 23-Jul-18 13:45 TileSeqMut/yogiroc_embeded.R
+-rw-rw-r--  2.0 unx     5470 b- defN 23-Jul-18 13:45 TileSeqMut/data/phix.fasta
+-rwxrwxr-x  2.0 unx     2579 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.data/scripts/makePRC
+-rwxrwxr-x  2.0 unx     1198 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.data/scripts/mergeRuns
+-rwxrwxr-x  2.0 unx      661 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.data/scripts/posterior_QC
+-rwxrwxr-x  2.0 unx      815 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.data/scripts/random_ds
+-rwxrwxr-x  2.0 unx     4219 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.data/scripts/tileseq_mut
+-rw-rw-r--  2.0 unx     1048 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.dist-info/LICENSE
+-rw-rw-r--  2.0 unx    14346 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.dist-info/METADATA
+-rw-rw-r--  2.0 unx       92 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.dist-info/WHEEL
+-rw-rw-r--  2.0 unx       11 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.dist-info/top_level.txt
+?rw-rw-r--  2.0 unx     2199 b- defN 23-Aug-02 14:37 TileSeqMut-0.6.944.dist-info/RECORD
+26 files, 226987 bytes uncompressed, 59246 bytes compressed:  73.9%
```

## zipnote {}

```diff
@@ -42,38 +42,38 @@
 
 Filename: TileSeqMut/yogiroc_embeded.R
 Comment: 
 
 Filename: TileSeqMut/data/phix.fasta
 Comment: 
 
-Filename: TileSeqMut-0.6.943.data/scripts/makePRC
+Filename: TileSeqMut-0.6.944.data/scripts/makePRC
 Comment: 
 
-Filename: TileSeqMut-0.6.943.data/scripts/mergeRuns
+Filename: TileSeqMut-0.6.944.data/scripts/mergeRuns
 Comment: 
 
-Filename: TileSeqMut-0.6.943.data/scripts/posterior_QC
+Filename: TileSeqMut-0.6.944.data/scripts/posterior_QC
 Comment: 
 
-Filename: TileSeqMut-0.6.943.data/scripts/random_ds
+Filename: TileSeqMut-0.6.944.data/scripts/random_ds
 Comment: 
 
-Filename: TileSeqMut-0.6.943.data/scripts/tileseq_mut
+Filename: TileSeqMut-0.6.944.data/scripts/tileseq_mut
 Comment: 
 
-Filename: TileSeqMut-0.6.943.dist-info/LICENSE
+Filename: TileSeqMut-0.6.944.dist-info/LICENSE
 Comment: 
 
-Filename: TileSeqMut-0.6.943.dist-info/METADATA
+Filename: TileSeqMut-0.6.944.dist-info/METADATA
 Comment: 
 
-Filename: TileSeqMut-0.6.943.dist-info/WHEEL
+Filename: TileSeqMut-0.6.944.dist-info/WHEEL
 Comment: 
 
-Filename: TileSeqMut-0.6.943.dist-info/top_level.txt
+Filename: TileSeqMut-0.6.944.dist-info/top_level.txt
 Comment: 
 
-Filename: TileSeqMut-0.6.943.dist-info/RECORD
+Filename: TileSeqMut-0.6.944.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## TileSeqMut/cluster.py

```diff
@@ -40,15 +40,15 @@
         os.system(sub_cmd)
         shfile_ds = os.path.join(sh_output, sample_name+"_aln_ds.sh")
         log_file_ds = alignment.align_main(ref, row["r1_ds"], row["r2_ds"], ds_sam_path, shfile_ds)
         sub_cmd = f"qsub -cwd -N {'aln_ds_'+sample_name} -e {log_file_ds} {shfile_ds}"
         os.system(sub_cmd)
 
 
-def alignment_sh_galen(fastq_map, ref_name, ref_seq, ref_path, sam_path, sh_output, at, logging, rc):
+def alignment_sh_galen(fastq_map, ref_name, ref_seq, ref_path, sam_path, sh_output, at, logging, rc, blacklist):
     """
     fastq_map: df contains paths to fastq files and downsamled fastq files
     ref_name: name for the reference sequence (same as project name)
     ref_seq: reference sequence to put in fasta file
     ref_path: build fasta index in this path
     sam_path: path to sam files
     sh_output: make sh files to submit to the cluster
@@ -60,14 +60,19 @@
     make SGE sh files
     submit sh files to SGE
     """
     ## build references
     ref = alignment.make_ref(ref_name, ref_seq, ref_path)
     phix = alignment.make_ref("phix", ref_seq, ref_path)
 
+    if blacklist != "": 
+        blacklistArg = f"#SBATCH --exclude={blacklist}\n"
+    else:
+        blacklistArg = ""
+
     # store sam paths
     fastq_map = pd.concat([fastq_map, pd.DataFrame(columns=["r1_sam", "r2_sam"])])
     all_job_id = []
     for index, row in fastq_map.iterrows():
         sample_name = os.path.basename(row["R1"]).split("_")[0]
 
         # for each sample, the alignment is for both R1 and R2 (they are aligned separately)
@@ -76,23 +81,23 @@
         # write header to sh file
 
         # create log file for alignment
         sam_log_f = os.path.join(sam_path, f"{sample_name}")
 
         if "Undetermined" in sample_name: # phix takes longer to align
             time_request = f"36:00:00"
-            header = f"#!/bin/bash\n#SBATCH --time={time_request}\n#SBATCH --mem=5G\n#SBATCH --job-name={sample_name}\n#SBATCH " \
-                 f"--error={sam_log_f}-%j.log\n#SBATCH --output={sam_log_f}-%j.log\n"
+            header = f"#!/bin/bash\n#SBATCH --time={time_request}\n#SBATCH --mem=2G\n#SBATCH --job-name={sample_name}\n#SBATCH " \
+                 f"--error={sam_log_f}-%j.log\n{blacklistArg}#SBATCH --output={sam_log_f}-%j.log\n"
             # when align undetermined fastq files to phix, we consider reads in both direction, rc = True
             r1_sam, r2_sam, log_file = alignment.align_main(phix, row["R1"], row["R2"], sam_path, shfile, rc=True,
                                                             header=header)
         else:
             time_request = f"0{at}:00:00"
-            header = f"#!/bin/bash\n#SBATCH --time={time_request}\n#SBATCH --mem=5G\n#SBATCH --job-name={sample_name}\n#SBATCH " \
-                 f"--error={sam_log_f}-%j.log\n#SBATCH --output={sam_log_f}-%j.log\n"
+            header = f"#!/bin/bash\n#SBATCH --time={time_request}\n#SBATCH --mem=2G\n#SBATCH --job-name={sample_name}\n#SBATCH " \
+                 f"--error={sam_log_f}-%j.log\n{blacklistArg}#SBATCH --output={sam_log_f}-%j.log\n"
             r1_sam, r2_sam, log_file = alignment.align_main(ref, row["R1"], row["R2"], sam_path, shfile, rc=rc, header=header)
 
         row["r1_sam"] = r1_sam
         row["r2_sam"] = r2_sam
         sub_cmd = ["sbatch", str(shfile)]
         jobs = subprocess.run(sub_cmd, stdout=subprocess.PIPE)
         job_id = jobs.stdout.decode("utf-8").strip().split(".")[0]
@@ -163,27 +168,33 @@
     job = subprocess.run(sub_cmd, stdout=subprocess.PIPE)
     job_id = job.stdout.decode("utf-8").strip().split(".")[0]
     # log sample name and job id
     logger.info(f"Sample {sample_name}: job id - {job_id}")
     return job_id
 
 
-def mut_count_sh_galen(sample_name, cmd, mt, mm, sh_output_dir, logger, cores):
+def mut_count_sh_galen(sample_name, cmd, mt, mm, sh_output_dir, logger, cores, blacklist):
     """
     Submit mutation count jobs to DC
     """
     # go through files df and submit jobs for each pair of sam files
     # counting mutations in raw sam output files
     shfile = os.path.join(sh_output_dir, f"Mut_count_{sample_name}.sh")
     log_f = os.path.join(sh_output_dir, f"Mut_count_{sample_name}")
     time_request = f"{mt}:00:00"
 
-    header = f"#!/bin/bash\n#SBATCH --time={time_request}\n#SBATCH --job-name={sample_name}\n#SBATCH " \
+    if blacklist != "": 
+        blacklistArg = f"#SBATCH --exclude={blacklist}\n"
+    else:
+        blacklistArg = ""
+
+
+    header = f"#!/bin/bash\n#SBATCH --time={time_request}\n#SBATCH --job-name=PTM{sample_name}\n#SBATCH " \
              f"--cpus-per-task={cores}\n#SBATCH --error={log_f}-%j.log\n#SBATCH --mem={mm}G\n#SBATCH " \
-             f"--output={log_f}-%j.log\n"
+             f"--output={log_f}-%j.log\n{blacklistArg}"
 
     with open(shfile, "w") as sh:
         sh.write(header)
         sh.write(cmd+"\n")
         os.system(f"chmod 755 {shfile}")
     #sample_error_file = os.path.join(log_dir, f"sample_{sample_name}.log")
     # submit this to the cluster
```

## TileSeqMut/count_mut.py

```diff
@@ -55,23 +55,23 @@
         # ending position of this tile (cds position)
         self._tile_ends = self._tile_map[self._tile_map["Tile Number"] == self._sample_tile]["End AA"].values[0] *3
         self._tile_len = self._tile_ends - self._tile_begins
         # min cover is the minimum percentage of the tile that needs to be covered by a read
         # for example, mincover=0.9 means 90 percent of the tile needs to be covered
         min_cover = self._var["minCover"]
         self._min_map_len = math.ceil(self._tile_len * float(min_cover))
-        self._cds_start = self._seq.cds_start
+        self._cds_start = self._seq.cds_start.iloc[0]
 
         # get sample specific info
         self._sample_condition = self._sample_info["Condition"].values[0]
         self._sample_tp = self._sample_info["Time point"].values[0]
         self._sample_rep = int(self._sample_info["Replicate"].values[0])
         # seq_lookup is a dataframe with the following columns:
         # temp_pos, cds, dna_pos, protein, protein_pos
-        self._seq_lookup = help_functions.build_lookup(self._cds_start.values.item(), self._seq.cds_end.values.item(), self._cds_seq)
+        self._seq_lookup = help_functions.build_lookup(self._cds_start, self._seq.cds_end.iloc[0], self._cds_seq)
 
         # path for the output files
         self._sample_counts_f = os.path.join(self._output_counts_dir,f"counts_sample_{self._sample_id}.csv")
         # r1 and r2 files are only used when posteriorQC is enabled
         self._sample_counts_r1_f = os.path.join(self._output_counts_dir, f"counts_sample_{self._sample_id}_r1.csv")
         self._sample_counts_r2_f = os.path.join(self._output_counts_dir, f"counts_sample_{self._sample_id}_r2.csv")
         # file used to track reads mapped outside of the tile
```

## TileSeqMut/help_functions.py

```diff
@@ -22,15 +22,15 @@
     return information about the run
     """
     with open(json_file, "r") as jsonf:
         data = json.load(jsonf)
         project = data["project"] # project name
         seq = pd.DataFrame(data["template"], index=[0]) # sequence
         cds_seq = seq.seq[0]
-        cds_seq = cds_seq[int(seq.cds_start)-1: int(seq.cds_end)]
+        cds_seq = cds_seq[int(seq.cds_start.iloc[0])-1: int(seq.cds_end.iloc[0])]
         # create dictionary to map tile/region # to start,end positions
         try:
             tile_map = pd.DataFrame(data["tiles"])
         except:
             tile_map = pd.DataFrame(data["tiles"], index=[0])
 
         try:
```

## TileSeqMut/main.py

```diff
@@ -184,15 +184,15 @@
             sh_output = os.path.join(self._output, "GALEN_jobs")
             os.system("mkdir "+sh_output)
             # for each pair of fastq files, submit jobs to run alignment
             # it takes the following arguments: R1, R2, ref, Output sam
             # the output log (bowtie log) would be in the same dir
             logging.info("Writing sh files for alignment (GALEN)")
             sam_df, job_list = cluster.alignment_sh_galen(fastq_map, self._project, self._seq.seq.values.item(), ref_path,
-                                            sam_output, sh_output, self._args.at, self._log, rc)
+                                            sam_output, sh_output, self._args.at, self._log, rc, self._args.blacklist)
             logging.info("Alignment jobs are submitted to GALEN..")
 
         elif self._args.environment == "BC2" or self._args.environment == "BC" or self._args.environment == "DC":
             # make sh files to submit to BC
             sh_output = os.path.join(self._output, "CCBR_sh")
             os.system("mkdir " + sh_output)
             # make sh files to submit to BC
@@ -349,15 +349,15 @@
         #     # (sample_name, cmd, mt, sh_output_dir, logger)
         #     job_id = cluster.mut_count_sh_dc(sample, cmd, self._args.mt, self._args.mm, sh_output, self._log,
         #                                      self._args.c)
         elif self._args.environment == "GALEN":
             logging.info("Submitting mutation counts jobs to GALEN...")
             # (sample_name, cmd, mt, sh_output_dir, logger)
             job_id = cluster.mut_count_sh_galen(sample, cmd, self._args.mt, self._args.mm, sh_output, self._log,
-                                                self._args.c)  # this
+                                                self._args.c, self._args.blacklist)  # this
         else:
             raise ValueError("Wrong environment")
 
         return job_id
 
     def _checkoutput(self, f):
         """
```

## Comparing `TileSeqMut-0.6.943.data/scripts/makePRC` & `TileSeqMut-0.6.944.data/scripts/makePRC`

 * *Files identical despite different names*

## Comparing `TileSeqMut-0.6.943.data/scripts/mergeRuns` & `TileSeqMut-0.6.944.data/scripts/mergeRuns`

 * *Files identical despite different names*

## Comparing `TileSeqMut-0.6.943.data/scripts/posterior_QC` & `TileSeqMut-0.6.944.data/scripts/posterior_QC`

 * *Files identical despite different names*

## Comparing `TileSeqMut-0.6.943.data/scripts/random_ds` & `TileSeqMut-0.6.944.data/scripts/random_ds`

 * *Files identical despite different names*

## Comparing `TileSeqMut-0.6.943.data/scripts/tileseq_mut` & `TileSeqMut-0.6.944.data/scripts/tileseq_mut`

 * *Files 2% similar despite different names*

```diff
@@ -52,13 +52,14 @@
                                                                "turn on this option will treat EVERYTHING as wt (not recommended). "
                                                                "Phred scores will be adjusted based on the first replicate")
 parser.add_argument("--calibratePhredWT", action="store_true", help="When this parameter is provided, use wt to calibrate phred scores")
 parser.add_argument("--calibratePhredPhix", action="store_true", help="When this parameter is provided, use phix alignments to calibrate phred scores")
 parser.add_argument("--errorOverride", action="store_true", help="When this parameter is provided, use the observed error probability instead of max(PHRED specification, observed error)")
 parser.add_argument("--resubmit", action="store_true", help="For a finished run, batch resubmit failed scripts (if any)")
 parser.add_argument("--logger_time", help="Time (in seconds) to wait before checking jobs status", default=300)
+parser.add_argument("--blacklist", help="Comma-separated list of cluster nodes to exclude from submissions", default="")
 parser.add_argument("--version", help="print version and exit")
 args = parser.parse_args()
 # print("TileSeqMut release 0.2.1: Supports multiprocess")
 
 print(f"TileSeqMut release {v}")
 main(args, v)
```

## Comparing `TileSeqMut-0.6.943.dist-info/LICENSE` & `TileSeqMut-0.6.944.dist-info/LICENSE`

 * *Files identical despite different names*

## Comparing `TileSeqMut-0.6.943.dist-info/METADATA` & `TileSeqMut-0.6.944.dist-info/METADATA`

 * *Files 2% similar despite different names*

```diff
@@ -1,30 +1,31 @@
 Metadata-Version: 2.1
 Name: TileSeqMut
-Version: 0.6.943
+Version: 0.6.944
 Summary: Analysis scriptsTileSeqMut for TileSeq sequencing data
-Home-page: https://github.com/RyogaLi/tilseq_mutcount
+Home-page: https://github.com/rothlab/tileseq_mutcount
 Author: ROUJIA LI
 Author-email: roujia.li@mail.utoronto.ca
+License: UNKNOWN
+Platform: UNKNOWN
 Classifier: Development Status :: 3 - Alpha
 Classifier: Intended Audience :: Developers
 Classifier: Topic :: Software Development :: Build Tools
 Classifier: License :: OSI Approved :: MIT License
 Classifier: Programming Language :: Python :: 3.7
 Classifier: Programming Language :: Python :: 3.8
 Requires-Python: >=3.7
 Description-Content-Type: text/markdown
-License-File: LICENSE
-Requires-Dist: pandas
-Requires-Dist: seaborn
-Requires-Dist: numpy
 Requires-Dist: biopython
 Requires-Dist: fpdf
-Requires-Dist: progressbar
 Requires-Dist: matplotlib
+Requires-Dist: numpy
+Requires-Dist: pandas
+Requires-Dist: progressbar
+Requires-Dist: seaborn
 
 ## TileSeq mutation count package
 
 This package is made to parse input sequecning files (fastq) with user provided parameter.csv file.
 Output of this pipeline is mutation counts for each pair of fastq files.
 
 ## Dependencies
@@ -130,26 +131,26 @@
 ```
 
 **Start the run**
 
 * Once the run starts, it will first submit alignment jobs to the cluster and keep tracking of all the submitted
  alignment jobs. Once all the jobs are finished, the pipeline will automatically submit another batch of jobs for
   mutation calling.
- 
+
  * if you want to skip alignment and only do mutation calls for existing sam files you can run the following command:
- 
+
 * **Example of skipping alignment:**
 
 ```
 tileseq_mut -p ~/dev/tilseq_mutcount/190506_param_MTHFR.csv -o /home/rothlab1/rli/dev/tilseq_mutcount/output
 /190506_MTHFR_WT_2020-01-29-17-07-04/ --skip_alignment -n rerun_mut_count
 ```
 
  * if you want to resubmit mutation calls for failed samples:
- 
+
 * **Example of resubmit:**
 
 ```
 tileseq_mut -p ~/dev/tilseq_mutcount/190506_param_MTHFR.csv -o /home/rothlab1/rli/dev/tilseq_mutcount/output
 /190506_MTHFR_WT_2020-01-29-17-07-04/190506_MTHFR_WT_2020-01-29-17-07-04_mut_count/ --resubmit -n resub
 ```
 
@@ -178,32 +179,32 @@
 `./ref/` - Reference fasta file and bowtie2 index
 
 `./env_jobs/` - Bash scripts for submitting the alignment jobs. If any of the sam file(s) didn't generate correctly, use `sbatch *.sh` to resubmit job for this sample
 
 `./sam_files/` - Alignment output (SAM) and log files for bowtie2. Also contains alignment log file. Use this to check alignment rate.
 
 `./name_time-stamped_mut_count/` - Mutation counts in each sample are saved in csv files
-    
+
     - `./count_sample_*.csv` - Raw mutation counts for each sample. With meta data in header. Variants are represented in hgvs format
-  
+
     - `./offreads_samplename.csv` - Contains information for all the reads that did not cover 90% of the tile
 
     - `./env_jobs/` - Bash scripts for summitting the mutation count jobs, also log files for each sample. 
-    
+
         * `./env_jobs/*.log` - log file from the cluster
         * `./env_jobs/*.sh`  - submission script
-    
+
     - `coverage_sample_name.csv` - File contains read counts for each position in the tile. There are
      four columns in the file: 
      pos: nt position of the tile 
      m_both: Number of variants found on both reads covering the site 
      m_r1: Number of variants found only on Read 1
      m_r2: Number of variants found only on Read 2
      passed: Number of variants passed filter
-    
+
     - `*_R1/R2_calibrate_phred.csv` - Calibrated Phred scores (if applicable)
 
 The count_sample_\*\*.csv is passed to tileseqMave for further analysis. If, for some reason, some of the count_sample_*.csv file did not generate correctly, you can resubmmit the job with the argument `--resubmit` (see example above)
 
 ### Alignment
 ---
 
@@ -338,7 +339,9 @@
                         Output folder
   --covOverride         Ignore coverage files, please use this if there's no coverage_* file for either of the run
   --subtractWT          Subtract wt counts from non-select before merging
   -log LOG_LEVEL, --log_level LOG_LEVEL
                         set log level: debug, info, warning, error, critical.
                         (default = debug)
 ```
+
+
```

## Comparing `TileSeqMut-0.6.943.dist-info/RECORD` & `TileSeqMut-0.6.944.dist-info/RECORD`

 * *Files 20% similar despite different names*

```diff
@@ -1,26 +1,26 @@
 TileSeqMut/__init__.py,sha256=G466q90MUSnR04e6jiShKjqf-yOf2AHIokn1dqXGDQY,178
 TileSeqMut/__main__.py,sha256=EN7h6ShSGYjKqKCFBqVLZ-1ZwLvn27MbVTgDe5yq_rs,119
 TileSeqMut/alignment.py,sha256=0l7wQtdJhuMCYAlWorEQy-LrSE-vUHTOYjFLdzCl8RM,2094
-TileSeqMut/cluster.py,sha256=ynlbRt9I57yQI9k6_zAMRr-_-zUC9ZUELJMlW2Vw-rw,15800
-TileSeqMut/count_mut.py,sha256=96C6VmkxJHyH7I-4mIO_QUig9MpZrM_FZTHXtAHxq4s,27653
-TileSeqMut/help_functions.py,sha256=iOX2eq74oLKahTkIsiEZ5SsxIpCxMnki2U4fJBM1534,4208
+TileSeqMut/cluster.py,sha256=O_SPpmd9YSq4GFotyb83e87nOPHuTjYOF9P3H3VJ3Xs,16108
+TileSeqMut/count_mut.py,sha256=Zy6GXZ7OmekXK95RWNDAKQPhzlJ3eYpzyBqNyVbSVKI,27641
+TileSeqMut/help_functions.py,sha256=fM7QMoAvRvxIlw8CjqAxU3fmcO7_lPciV4_Ktz4Tycc,4224
 TileSeqMut/locate_mut.py,sha256=Ohgrj8LpI2F8TwpmOL7e8wc8QlwdSdWAjlpMbDtkV6A,24363
-TileSeqMut/main.py,sha256=5subJUrrHBxCLvRgsh-T90fVJws_4wmhLeGaSxlZIDI,33636
+TileSeqMut/main.py,sha256=UGEhpNUBPQ6ifP-UM8f13JT94_dXozMObiaDdbJ3Ng8,33680
 TileSeqMut/mergeruns.py,sha256=_WfIq9HUhnZL0hBt9ZBZCN7WBYktOVz-u5Cae5BUr1s,18853
 TileSeqMut/parse_clinvar.py,sha256=dT7xJQhiWhatOHbjSGPo8sT0tnCfKG7_Twe0nDfHJlI,17020
 TileSeqMut/posterior.py,sha256=i1ckY8b6CdaQvFeMZovWBDHe6kiPgzgafrYFg_jK28U,21518
 TileSeqMut/posterior_qc.py,sha256=CG1aLIpmyL7O0gbRuHqb13EXNF-RXbXEA9am9MpWsds,22185
 TileSeqMut/process.py,sha256=m2hKw0AvNG0N77bAGJOzMKsJ064sytmHM8bTBMwamiA,1546
 TileSeqMut/random_ds.py,sha256=pnJIO2-2XD2ltbvsCkcZXHV6inGlms356IzqllmGexI,3073
 TileSeqMut/yogiroc_embeded.R,sha256=Bn5phiifUoM9ZnOMDTi_UWAMkSypOdy8HQXnp5j9Wsg,1747
 TileSeqMut/data/phix.fasta,sha256=-XBxFcpVntYE1maG_ah9IXvXCRSkvrC2KfcLPG4Qjj0,5470
-TileSeqMut-0.6.943.data/scripts/makePRC,sha256=FEGHHpWJkPGVgEBgK2oxoEHC4OGdDo6-hqeJEidYxH4,2579
-TileSeqMut-0.6.943.data/scripts/mergeRuns,sha256=OCvseMWFjDYkhZKYcZ8qIOSXSaMtO8QX78NXPDvRkp8,1198
-TileSeqMut-0.6.943.data/scripts/posterior_QC,sha256=dp5Lmowp4Zm4Qto37rmESkZ-gzpS92oGAI5O7dpEZ5U,661
-TileSeqMut-0.6.943.data/scripts/random_ds,sha256=et5PDrFs16Kwfbn4pCJNcCD5mdumwdrGBaB-gxkTxvQ,815
-TileSeqMut-0.6.943.data/scripts/tileseq_mut,sha256=RefzM0VYXAfWMeGQaut37neP57nhoIeW-94H1tE37Zg,4098
-TileSeqMut-0.6.943.dist-info/LICENSE,sha256=rV02maV8WjM_a-dUbS49IYhv1ISXN549JbDlOPzya9o,1048
-TileSeqMut-0.6.943.dist-info/METADATA,sha256=-jV7mbDFu4JJutXivuQXEbhz23I26ANdXeVr4z_3-LY,14351
-TileSeqMut-0.6.943.dist-info/WHEEL,sha256=2wepM1nk4DS4eFpYrW1TTqPcoGNfHhhO_i5m4cOimbo,92
-TileSeqMut-0.6.943.dist-info/top_level.txt,sha256=QDiT9SnGqeyGN7kBWokoeYz_DxPfg46mEa41XoTecDs,11
-TileSeqMut-0.6.943.dist-info/RECORD,,
+TileSeqMut-0.6.944.data/scripts/makePRC,sha256=FEGHHpWJkPGVgEBgK2oxoEHC4OGdDo6-hqeJEidYxH4,2579
+TileSeqMut-0.6.944.data/scripts/mergeRuns,sha256=OCvseMWFjDYkhZKYcZ8qIOSXSaMtO8QX78NXPDvRkp8,1198
+TileSeqMut-0.6.944.data/scripts/posterior_QC,sha256=dp5Lmowp4Zm4Qto37rmESkZ-gzpS92oGAI5O7dpEZ5U,661
+TileSeqMut-0.6.944.data/scripts/random_ds,sha256=et5PDrFs16Kwfbn4pCJNcCD5mdumwdrGBaB-gxkTxvQ,815
+TileSeqMut-0.6.944.data/scripts/tileseq_mut,sha256=ISMsq4D0R-HTm3Um0szOYvzixjGX2Q7DHncFji1AaYQ,4219
+TileSeqMut-0.6.944.dist-info/LICENSE,sha256=rV02maV8WjM_a-dUbS49IYhv1ISXN549JbDlOPzya9o,1048
+TileSeqMut-0.6.944.dist-info/METADATA,sha256=xdt0X7bIl10LqMx7YKU8ruuO881fjASRNyOr91UEteE,14346
+TileSeqMut-0.6.944.dist-info/WHEEL,sha256=g4nMs7d-Xl9-xC9XovUrsDHGXt-FT0E17Yqo92DEfvY,92
+TileSeqMut-0.6.944.dist-info/top_level.txt,sha256=QDiT9SnGqeyGN7kBWokoeYz_DxPfg46mEa41XoTecDs,11
+TileSeqMut-0.6.944.dist-info/RECORD,,
```

